src/management-system-v2/lib/data/db/iam/users.ts

```diff
@@ -0,0 +1,175 @@
+  const user = await db.user.findUnique({ where: { username } });
+
+  if (!user && opts?.throwIfNotFound) throw new Error('User not found');
+
+  return user as User;
+}
+
+export async function addUser(inputUser: OptionalKeys<User, 'id'>) {
+  const user = UserSchema.parse(inputUser);
+
+  if (
+    !user.isGuest &&
+    ((user.username && (await getUserByUsername(user.username))) ||
+      (await getUserByEmail(user.email!)))
+  )
+    throw new Error('User with this email or username already exists');
+
+  if (!user.id) user.id = v4();
+
+  try {
+    const userExists = await db.user.findUnique({ where: { id: user.id } });
+    if (userExists) throw new Error('User already exists');
+    await db.user.create({
+      data: {
+        ...user,
+        isGuest: user.isGuest,
+      },
+    });
+    await addEnvironment({ ownerId: user.id, isOrganization: false });
+  } catch (error) {
+    console.error('Error adding new user: ', error);
+  }
+
+  if ((await getSystemAdmins()).length === 0)
+    addSystemAdmin({
+      role: 'admin',
+      userId: user.id,
+    });
+
+  return user as User;
+}
+
+export class UserHasToDeleteOrganizationsError extends Error {
+  conflictingOrgs: string[];
+
+  constructor(conflictingOrgs: string[], message?: string) {
+    super(message ?? 'User has to delete organizations before being deleted');
+    this.name = 'UserHasToDeleteOrganizationsError';
+    this.conflictingOrgs = conflictingOrgs;
+  }
+}
+export async function deleteUser(userId: string) {
+  const user = await db.user.findUnique({ where: { id: userId } });
+  if (!user) throw new Error("User doesn't exist");
+
+  const userOrganizations = await getUserOrganizationEnvironments(userId);
+  const orgsWithNoNextAdmin: string[] = [];
+  for (const environmentId of userOrganizations) {
+    const userRoles = await getRoleMappingByUserId(userId, environmentId);
+
+    if (!userRoles.find((role) => role.roleName === '@admin')) continue;
+
+    const adminRole = await db.role.findFirst({
+      where: { name: '@admin', environmentId: environmentId },
+      include: { members: true },
+    });
+    if (!adminRole)
+      throw new Error(`Consistency error: admin role of environment ${environmentId} not found`);
+
+    if (adminRole.members.length === 1) orgsWithNoNextAdmin.push(environmentId);
+  }
+
+  if (orgsWithNoNextAdmin.length > 0)
+    throw new UserHasToDeleteOrganizationsError(orgsWithNoNextAdmin);
+
+  await db.space.deleteMany({ where: { ownerId: userId } });
+  await db.user.delete({ where: { id: userId } });
+
+  return user;
+}
+
+export async function updateUser(userId: string, inputUser: Partial<AuthenticatedUser>) {
+  const user = await getUserById(userId, { throwIfNotFound: true });
+  const isGoingToBeGuest = inputUser.isGuest !== undefined ? inputUser.isGuest : user?.isGuest;
+
+  let updatedUser;
+  if (isGoingToBeGuest) {
+    updatedUser = { isGuest: true };
```

Comment:
added return updatedUserFromDB

Reply:
I meant the parameter `inputUser` is ignored if this flag is true. Only `updateUser` will be used and that only contains this field and ignores `inputUser`.

Reply:
yes, in general the guest user is not allowed to update any user info.

the updateUser function is called twice, in following scenario

1. When the guest user wants to sign in, at that moment, isGuest: false is set in db, rest of the params like username..etc are undefined and email is sent to the guest,
2. When the user verifies itself using the link sent via email, then the modal is shown to the user to update other fields username, first name ..,

Reply:
I could see some future use though, something like a promotional discount marker they can use when upgrading to a real account and that is set on the guest user if they register over a certain link or sth. like that.

My main concern is that in the future someone might not look this function up and assume it updates the user but would just ignore the input. If we want to hardcode that guest user can't change any other fields, than maybe this function should throw if any other fields are set instead of silently ignore it.

Reply:
I have adjusted the function to throw an err if a guest user tries to update user info for now.